src/bin/sccache-dist/main.rs

```diff
@@ -332,6 +371,62 @@ fn run(command: Command) -> Result<i32> {
+                    let kid_to_pkcs1: HashMap<String, Vec<u8>> = jwks.keys.into_iter().map(|k| {
+                        let pkcs1 = k.to_der_pkcs1().unwrap();
+                        (k.kid, pkcs1)
+                    }).collect();
+                    Box::new(move |s| {
+                        let header = jwt::decode_header(s).unwrap();
+                        // Prepare validation
+                        let kid = header.kid.unwrap();
+                        let pkcs1 = kid_to_pkcs1.get(&kid).unwrap();
+                        let mut validation = jwt::Validation::new(header.alg);
+                        validation.set_audience(&audience);
+                        validation.iss = Some(issuer.clone());
+                        #[derive(Deserialize)]
+                        struct Claims {}
+                        // Decode the JWT, discarding any claims - we just care about validity
+                        let _tokendata = jwt::decode::<Claims>(s, pkcs1, &validation).unwrap();
+                        true
+                    })
+                },
+                scheduler_config::ClientAuth::ProxyToken { url, cache_secs } => {
+                    let maybe_auth_cache: Option<Mutex<(HashMap<String, Instant>, Duration)>> =
+                        cache_secs.map(|secs| Mutex::new((HashMap::new(), Duration::from_secs(secs))));
+                    let client = reqwest::Client::new();
+                    Box::new(move |s| {
+                        // If the token is cached and not expired, return it
+                        if let Some(ref auth_cache) = maybe_auth_cache {
+                            let mut auth_cache = auth_cache.lock().unwrap();
+                            let (ref mut auth_cache, cache_duration) = *auth_cache;
+                            if let Some(cached_at) = auth_cache.get(s) {
+                                if cached_at.elapsed() < cache_duration {
+                                    return true
+                                }
+                            }
+                            auth_cache.remove(s);
+                        }
+                        // Make a request to check for token validity
+                        let header = reqwest::header::Authorization(reqwest::header::Bearer { token: s.to_owned() });
+                        let res = client.get(&url).header(header).send().unwrap();
+                        if !res.status().is_success() {
+                            return false
+                        }
+                        // Cache the token
+                        if let Some(ref auth_cache) = maybe_auth_cache {
+                            let mut auth_cache = auth_cache.lock().unwrap();
+                            let (ref mut auth_cache, _) = *auth_cache;
+                            auth_cache.insert(s.to_owned(), Instant::now());
+                        }
+                        true
+                    })
```

Comment:
These two new match blocks are rather large. Maybe they'd be better off factored out into separate functions?